src/prefect/engine/results/json_result.py

```diff
@@ -0,0 +1,57 @@
+import json
+
+from typing import Any
+
+from prefect.engine.result import Result
+
+
+class JSONResult(Result):
+    """
+    Hook for storing and retrieving JSON serializable Python objects that can
+    safely be stored directly in a Prefect database.
+
+    Args:
+        - value (Any): the underlying value this Result should represent
+    """
+
+    def __init__(self, value: Any = None, **kwargs: Any) -> None:
+        self.value = value
+        super().__init__(value=value, **kwargs)
+
+    def read(self, loc: str = None) -> Result:
+        """
+        Returns the underlying value regardless of the argument passed.
+
+        Args:
+            - loc (str): an unused argument
+        """
+        new = self.copy()
+        new.value = json.loads(loc or new.filepath)
+        return new
+
+    def write(self, **kwargs: Any) -> Result:
+        """
+        Returns the repr of the underlying value, purely for convenience.
+
+        Args:
+            - **kwargs (optional): unused, for compatibility with the interface
+
+        Returns:
+            - Result: returns self
+        """
+        self.filepath = json.dumps(self.value)
+        return self
+
+    def exists(self, loc: str = None) -> bool:
```

Comment:
Is this the same as Constant result on purpose? Unlike Constants which we expect to have a value at instantiation, it seems to me a JSONResult shouldn't consider itself to `exist` until the task has run aka its value has been 'written' to its self.filepath.

Kind of some steps down the way still but are we going to have a problem distinguishing between JSONResults that have a None type value/filepath because their task hasn't been run, and JSONResults that have a None type value/filepath because that is the return value of their task? I think this is what NoResult type is about, and is probably hard to discuss (for me at least) without hooking in the exists pipeline checks more, but I started thinking about it while reading this.

Reply:
Hmmm that's a good call out; in practice the value is available for use if you're interacting with the class, so I'm tempted to keep it as-is (and I also wouldn't recommend anyone use this type for their `target`, which is where this method will be relied on)